onnx/shape_inference/implementation.h

```diff
@@ -241,6 +241,148 @@ struct InferenceContextImpl : public InferenceContext {
+    if (index >= allOutputTypes_.size()) {
+      ONNX_THROW("Output " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return &allOutputTypes_[index];
+  }
+
+  // Convert integer vector into TensorShapeProto
+  template <typename INTEGER>
+  TensorShapeProto vectorToTensorShapeProto(const std::vector<INTEGER>& input_vals) const {
+    TensorShapeProto converted_tsp;
+    for (unsigned int i = 0; i < input_vals.size(); ++i) {
+      converted_tsp.mutable_dim()->Add()->set_dim_value(input_vals[i]);
+    }
+    return converted_tsp;
+  }
+
+  const TensorShapeProto* getInputData(size_t index) override {
+    if (index >= allInputData_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    const std::string input_name = inputIndexToNameMap_.at(index);
+    // Gets it from previous data propagation
+    auto iter = generatedShapeData_.find(input_name);
+    if (iter != generatedShapeData_.end()) {
+      return &iter->second;
+    }
+    // Otherwise, gets it from initializer if it exists
+    const auto* input_data = allInputData_[index];
+    // Only scalar (0D tensor) or 1D tensor can be converted for now
+    // TODO: It should support tensors with more dimension on demand
+    if (input_data != nullptr &&
+        (input_data->dims_size() == 0 || input_data->dims_size() == 1)) {
+        TensorShapeProto tsp;
+
+        if (input_data->data_type() == TensorProto_DataType_INT64) {
+          tsp = vectorToTensorShapeProto(ParseData<int64_t>(input_data));
+        } else if (input_data->data_type() == TensorProto_DataType_INT32) {
+          tsp = vectorToTensorShapeProto(ParseData<int32_t>(input_data));
+        } else {
+          // Only supports integer type to form a shape
+          return nullptr;
+        }
+
+        // Adds this TensorShapeProto from initializer into generatedShapeData
+        // for future use
+        generatedShapeData_.insert({input_name, std::move(tsp)});
+        auto iter = generatedShapeData_.find(input_name);
+        return &iter->second;
+    }
+    return nullptr;
+  }
+
+  void addOutputData(size_t index, TensorShapeProto&& tsp) override {
+    if (index >= outputIndexToNameMap_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    auto result = generatedShapeData_.insert({outputIndexToNameMap_.at(index), std::move(tsp)});
+    if (!result.second) {
+      fail_shape_inference("Data for input  " + ONNX_NAMESPACE::to_string(index) + " already exists.");
+    }
+  }
+
+  std::vector<const TensorProto*> allInputData_;
+  std::unordered_map<size_t, std::string> inputIndexToNameMap_;
+  std::unordered_map<size_t, std::string> outputIndexToNameMap_;
+  std::vector<const TypeProto*> allInputTypes_;
+  std::vector<TypeProto> allOutputTypes_;
+  std::unordered_map<std::string, TensorShapeProto>& generatedShapeData_;
+  const GraphInferenceContext* graphInferenceContext_;
```

Comment:
will we even need GraphInferenceContext in DataPropagation?

Reply:
For now no, I think only control ops like `Loop` or `If` will need it and they are not included in the supported list for data propagation. I have removed it. Thanks.